### PR TITLE
Add internal annotation

### PR DESCRIPTION
**Pull request by giacomocavalieri**
This PR adds the `@internal` annotation, closing #2486 

### ~~Why is this a WIP PR?~~

- [x] I'm waiting for #2537 to be merged and exclude internals from the package interface as well

### CODE REVIEW

File: compiler-core/src/language_server/engine.rs [hunk 1]
```diff
@@ -368,6 +368,9 @@ where
 
         // Module types
         for (name, type_) in &module.ast.type_info.types {
+            if type_.internal {
+                continue;
+            }
```
Comment (DanielSherlock) [inline compiler-core/src/language_server/engine.rs:373]: Are these lines of code excluding `@internal` types from the same file from completions? I don't think we want that.

File: compiler-core/src/language_server/tests/completion.rs [hunk 2]
```diff
@@ -726,3 +726,71 @@ pub fn wibble(
...
+@internal pub type Alias = Int
+@internal pub type Foo { Bar }
+@internal pub const foo = 1
+"#;
+
+    assert_eq!(expression_completions("", code), vec![]);
+}
+
+#[test]
+fn public_internal_definitions_from_the_same_module_are_in_the_completions() {
```
Comment (DanielSherlock) [inline compiler-core/src/language_server/tests/completion.rs:747]: This test seems to only check value-completion, not type-completions. In particular the types `Foo` and `Alias`.

I think these would need a separate test (like `for_custom_type_definition()` at line 568 or the five following tests) which would then detect the I-think-incorrect behaviour in my above comment.

**Comment by DanielSherlock** [comment]
_(I don't rust, so feel free to ignore me completely.)_

I joined the stream late so was curious to look at parts of the code again after, and noticed some things that seem suspicious about completions for types. Hopefully the comments I added are getting at the right bits of code for the issue.

File: compiler-core/src/language_server/engine.rs (hunk 1 above)
Comment (giacomocavalieri) [reply compiler-core/src/language_server/engine.rs:373]: I think you're right! I will look into it
Thanks for pointing it out!!

Comment (giacomocavalieri) [reply compiler-core/src/language_server/engine.rs:373]: I checked and you're perfectly right! Thank you so much for the help 💜

**Comment by giacomocavalieri** [comment]
Thank you @DanielSherlock for pointing this out! I've fixed the bug and added some tests to make sure that internal types do appear in the completions for the same module they're defined in

**Comment by lpil** [comment]
Could you rebase on main please? 🙏

File: compiler-core/src/analyse.rs [hunk 3]
```diff
@@ -537,7 +547,14 @@ fn register_value_from_function(
...
     };
-    environment.insert_variable(name.clone(), variant, typ, *public, deprecation.clone());
+    environment.insert_variable(
+        name.clone(),
+        variant,
+        typ,
+        *public,
+        *internal,
+        deprecation.clone(),
+    );
```
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/analyse.rs:591]: Ok! Should it be a compile error or should it be a bit more permissive and the formatter could just remove the internal annotation on private functions?

File: compiler-core/src/language_server/engine.rs [hunk 4]
```diff
@@ -382,7 +385,7 @@ where
 
             // Qualified types
             for (name, type_) in &module.types {
-                if !type_.public {
+                if !type_.public || type_.internal {
```
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/language_server/engine.rs:388]: Wouldn't removing it show _all_ internal types in the autocompletions - even those defined by external dependencies?

[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/src/language_server/engine.rs:388]: Aye, so it needs to check the package too

File: compiler-core/src/analyse.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/src/analyse.rs:591]: Oh yes good point. Perhaps a compile error would be good?

**Comment by lpil** [comment]
Is this WIP @giacomocavalieri ?

**Comment by giacomocavalieri** [comment]
Yes I still have to finish addressing your review comments, I'll be working on this shortly!

File: compiler-core/src/language_server/engine.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/language_server/engine.rs:385]: I'm not sure how I can check if a module come from the root package, how can I do that?

File: compiler-core/src/analyse.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (giacomocavalieri) [reply compiler-core/src/analyse.rs]: Ok! I went with an error

**Comment by giacomocavalieri** [comment]
Turning `public` + `internal` into a single `publicity` attribute turned out to be quite the chunky refactoring but I think this is finally ready for review @lpil 🎉

File: compiler-core/src/docs.rs [hunk 5]
```diff
@@ -164,6 +167,7 @@ pub fn generate_html(
             .ast
             .definitions
             .iter()
+            .filter(|statement| !statement.is_internal())
```
Comment (lpil) [inline compiler-core/src/docs.rs:170]: nit: is_public() would be a wee bit clearer here

File: compiler-core/src/docs.rs [hunk 6]
```diff
@@ -559,7 +563,7 @@ fn type_<'a>(source_links: &SourceLinker, statement: &'a TypedDefinition) -> Opt
     let mut formatter = format::Formatter::new();
 
     match statement {
-        Definition::CustomType(ct) if ct.public && !ct.opaque => Some(Type {
+        Definition::CustomType(ct) if !ct.publicity.is_private() && !ct.opaque => Some(Type {
```
Comment (lpil) [inline compiler-core/src/docs.rs:566]: Same here: `.is_public()`

File: compiler-core/src/javascript/typescript.rs [hunk 7]
```diff
@@ -294,41 +294,43 @@ impl<'a> TypeScriptGenerator<'a> {
         match statement {
             Definition::TypeAlias(TypeAlias {
                 alias,
-                public,
+                publicity,
                 type_,
                 ..
-            }) if *public => vec![self.type_alias(alias, type_)],
+            }) if !publicity.is_internal() => vec![self.type_alias(alias, type_)],
```
Comment (lpil) [inline compiler-core/src/javascript/typescript.rs:300]: Why is this rendered for public and private but not internal?

File: compiler-core/src/language_server/engine.rs [hunk 8]
```diff
@@ -377,10 +388,13 @@ where
...
-                    continue;
-                };
-                completions.push(type_completion(None, unqualified.used_name(), type_));
+                match module.get_public_type(&unqualified.name) {
+                    Some(type_) if type_.publicity.is_internal() => continue,
+                    Some(type_) => {
+                        completions.push(type_completion(None, unqualified.used_name(), type_))
+                    }
+                    None => continue,
+                }
```
Comment (lpil) [inline compiler-core/src/language_server/engine.rs:397]: In this case they've been imported explicitly so I think we should show them.

File: compiler-core/src/language_server/engine.rs [hunk 9]
```diff
@@ -354,6 +355,8 @@ where
             completions.push(type_completion(None, name, type_));
         }
 
+        let root_package = self.compiler.project_compiler.config.name.clone();
```
Comment (lpil) [inline compiler-core/src/language_server/engine.rs:358]: I don't think we need to clone here, using a `&str` will be fine. A helper function `self.root_package_name()` could be good too

File: compiler-core/src/language_server/engine.rs [hunk 10]
```diff
@@ -392,9 +406,14 @@ where
 
         // Module functions
         for (name, value) in &module.ast.type_info.values {
+            // Here we do not check for the internal attribute: we always want
+            // to show autocompletions for values defined in the same module,
+            // even if those are internal.
             completions.push(value_completion(None, name, value));
         }
 
+        let root_package = self.compiler.project_compiler.config.name.clone();
```
Comment (lpil) [inline compiler-core/src/language_server/engine.rs:415]: Same as above

File: compiler-core/src/language_server/engine.rs [hunk 11]
```diff
@@ -418,10 +445,13 @@ where
 
             // Unqualified values
             for unqualified in &import.unqualified_values {
-                let Some(value) = module.get_public_value(&unqualified.name) else {
-                    continue;
-                };
-                completions.push(value_completion(None, unqualified.used_name(), value));
+                match module.get_public_value(&unqualified.name) {
+                    Some(value) if value.publicity.is_internal() => continue,
```
Comment (lpil) [inline compiler-core/src/language_server/engine.rs:449]: As above, they've been explicitly brought into scope so let's autocomplete them

File: compiler-core/src/language_server/tests/completion.rs [hunk 12]
```diff
@@ -19,13 +23,33 @@ fn positioned_expression_completions(
...
     dep: &str,
     position: Position,
+) -> Vec<CompletionItem> {
+    positioned_expression_completions_with_dep(src, dep, false, position)
+}
+
+/// `external` is used to tell whether the specified dependency has to be
+/// treated as coming from an external imported package or as coming from a
+/// module from the same root package.
+/// In any case `src` will be able to import it using `import dep`.
```
Comment (lpil) [inline compiler-core/src/language_server/tests/completion.rs:33]: External means it's written in FFI, dependency means it comes from another package. I think with the old name it should have always been in a dep package.

File: compiler-core/src/language_server/tests/completion.rs [hunk 13]
```diff
@@ -19,13 +23,33 @@ fn positioned_expression_completions(
...
+    if external {
+        add_path_dep(&mut engine, "dep");
+        let _ = io.path_dep_module("dep", "dep", dep);
+    } else {
+        _ = io.src_module("dep", dep);
+    }
     _ = io.src_module("app", src);
+
     let response = engine.compile_please();
+    println!("{:#?}", response);
```
Comment (lpil) [inline compiler-core/src/language_server/tests/completion.rs:52]: Printing here 😁

**Review by lpil — COMMENTED** [review]
Nice!! And what an effort 💪  I've left a few small notes inline.

File: compiler-core/src/language_server/tests/completion.rs (hunk 13 above)
Comment (giacomocavalieri) [reply compiler-core/src/language_server/tests/completion.rs:52]: 🫢

File: compiler-core/src/docs.rs (hunk 5 above)
Comment (giacomocavalieri) [reply compiler-core/src/docs.rs:170]: Aha the problem is that we want to do it with both public and internals, so anything that is not private, how would you call such a method?
```rust
fn wibble(&self) -> bool {
  case self {
    Publicity::Internal | Publicity::Public => true,
    Publicity::Private => false,
  }
}
```

File: compiler-core/src/javascript/typescript.rs (hunk 7 above)
Comment (giacomocavalieri) [reply compiler-core/src/javascript/typescript.rs:300]: Oh that's a typo! It was meant to be `!publicity.is_private()`

Comment (lpil) [reply compiler-core/src/javascript/typescript.rs:300]: oh no! 😁

File: compiler-core/src/docs.rs (hunk 5 above)
Comment (lpil) [reply compiler-core/src/docs.rs:170]: I realised this for the later ones 😓 

I wonder if `!x.is_internal()` should be `x.is_importable()` or something?

Comment (giacomocavalieri) [reply compiler-core/src/docs.rs:170]: Ah I really like that name, I'll go for it and we can always change it later 😁

File: compiler-core/src/language_server/tests/completion.rs (hunk 12 above)
Comment (giacomocavalieri) [reply compiler-core/src/language_server/tests/completion.rs:33]: The way it was implemented it was just considered as another module under the same root package. I can change it to make sure the dep is always under another package if you want

**Review by lpil — APPROVED** [review]
Thank you!!!

**Pull request merged**